File: sdks/cpp/src/permissions.cpp
#include "iora/permissions.hpp"
#include <unordered_map>

namespace iora {

std::string permission_to_string(Permission permission) {
    static const std::unordered_map<Permission, std::string> permission_strings = {
        {Permission::READ_ENTITIES, "ReadEntities"},
        {Permission::CONTROL_ENTITIES, "ControlEntities"},
        {Permission::CREATE_ENTITIES, "CreateEntities"},
        {Permission::DELETE_ENTITIES, "DeleteEntities"},
        {Permission::STORAGE_READ, "StorageRead"},
        {Permission::STORAGE_WRITE, "StorageWrite"},
        {Permission::STORAGE_DELETE, "StorageDelete"},
        {Permission::NETWORK_ACCESS, "NetworkAccess"},
        {Permission::NETWORK_OUTBOUND, "NetworkOutbound"},
        {Permission::NETWORK_INBOUND, "NetworkInbound"},
        {Permission::NETWORK_SCAN, "NetworkScan"},
        {Permission::NETWORK_LOCAL_ACCESS, "NetworkLocalAccess"},
        {Permission::SYSTEM_INFO, "SystemInfo"},
        {Permission::SYSTEM_CONTROL, "SystemControl"},
        {Permission::SYSTEM_RESTART, "SystemRestart"},
        {Permission::DATABASE_READ, "DatabaseRead"},
        {Permission::DATABASE_WRITE, "DatabaseWrite"},
        {Permission::DATABASE_CREATE, "DatabaseCreate"},
        {Permission::DATABASE_DELETE, "DatabaseDelete"},
        {Permission::REGISTER_API, "RegisterApi"},
        {Permission::CALL_API, "CallApi"},
        {Permission::REGISTER_WIDGET, "RegisterWidget"},
        {Permission::CONTROL_WIDGET, "ControlWidget"},
        {Permission::SEND_NOTIFICATIONS, "SendNotifications"},
        {Permission::READ_NOTIFICATIONS, "ReadNotifications"},
        {Permission::CAMERA_ACCESS, "CameraAccess"},
        {Permission::MICROPHONE_ACCESS, "MicrophoneAccess"},
        {Permission::MEDIA_ACCESS, "MediaAccess"},
        {Permission::PLUGIN_MANAGER, "PluginManager"},
        {Permission::INSTALL_PLUGINS, "InstallPlugins"},
        {Permission::FILE_SYSTEM_READ, "FileSystemRead"},
        {Permission::FILE_SYSTEM_WRITE, "FileSystemWrite"},
        {Permission::FILE_SYSTEM_EXECUTE, "FileSystemExecute"},
        {Permission::LOCATION_ACCESS, "LocationAccess"},
        {Permission::AUTOMATIONS, "Automations"}
    };

    auto it = permission_strings.find(permission);
    if (it != permission_strings.end()) {
        return it->second;
    }
    return "Unknown";
}

RiskLevel get_permission_risk_level(Permission permission) {
    switch (permission) {
        // Low risk
        case Permission::READ_ENTITIES:
        case Permission::STORAGE_READ:
        case Permission::DATABASE_READ:
        case Permission::SYSTEM_INFO:
        case Permission::READ_NOTIFICATIONS:
            return RiskLevel::LOW;

        // Critical risk
        case Permission::SYSTEM_CONTROL:
        case Permission::SYSTEM_RESTART:
        case Permission::PLUGIN_MANAGER:
        case Permission::INSTALL_PLUGINS:
        case Permission::FILE_SYSTEM_WRITE:
        case Permission::FILE_SYSTEM_EXECUTE:
        case Permission::NETWORK_SCAN:
        case Permission::NETWORK_LOCAL_ACCESS:
            return RiskLevel::CRITICAL;

        // High risk
        case Permission::CREATE_ENTITIES:
        case Permission::DELETE_ENTITIES:
        case Permission::NETWORK_ACCESS:
        case Permission::NETWORK_OUTBOUND:
        case Permission::REGISTER_API:
        case Permission::REGISTER_WIDGET:
        case Permission::CAMERA_ACCESS:
        case Permission::FILE_SYSTEM_READ:
            return RiskLevel::HIGH;

        // Medium risk (default)
        default:
            return RiskLevel::MEDIUM;
    }
}

std::string get_permission_description(Permission permission) {
    static const std::unordered_map<Permission, std::string> descriptions = {
        {Permission::READ_ENTITIES, "Read entity states and attributes"},
        {Permission::CONTROL_ENTITIES, "Control smart home devices"},
        {Permission::CREATE_ENTITIES, "Create new entities"},
        {Permission::DELETE_ENTITIES, "Delete entities"},
        {Permission::STORAGE_READ, "Read from storage"},
        {Permission::STORAGE_WRITE, "Write to storage"},
        {Permission::STORAGE_DELETE, "Delete from storage"},
        {Permission::NETWORK_ACCESS, "Access external networks"},
        {Permission::NETWORK_OUTBOUND, "Make outbound network requests"},
        {Permission::NETWORK_INBOUND, "Accept inbound network connections"},
        {Permission::NETWORK_SCAN, "Scan local network"},
        {Permission::NETWORK_LOCAL_ACCESS, "Access local network devices"},
        {Permission::SYSTEM_INFO, "Read system information"},
        {Permission::SYSTEM_CONTROL, "Control system settings"},
        {Permission::SYSTEM_RESTART, "Restart IORA system"},
        {Permission::DATABASE_READ, "Read from database"},
        {Permission::DATABASE_WRITE, "Write to database"},
        {Permission::DATABASE_CREATE, "Create database tables"},
        {Permission::DATABASE_DELETE, "Delete from database"},
        {Permission::REGISTER_API, "Register API endpoints"},
        {Permission::CALL_API, "Call API endpoints"},
        {Permission::REGISTER_WIDGET, "Register dashboard widgets"},
        {Permission::CONTROL_WIDGET, "Control widgets"},
        {Permission::SEND_NOTIFICATIONS, "Send notifications"},
        {Permission::READ_NOTIFICATIONS, "Read notifications"},
        {Permission::CAMERA_ACCESS, "Access camera"},
        {Permission::MICROPHONE_ACCESS, "Access microphone"},
        {Permission::MEDIA_ACCESS, "Access media files"},
        {Permission::PLUGIN_MANAGER, "Manage plugins"},
        {Permission::INSTALL_PLUGINS, "Install/uninstall plugins"},
        {Permission::FILE_SYSTEM_READ, "Read files"},
        {Permission::FILE_SYSTEM_WRITE, "Write files"},
        {Permission::FILE_SYSTEM_EXECUTE, "Execute files"},
        {Permission::LOCATION_ACCESS, "Access location data"},
        {Permission::AUTOMATIONS, "Create and manage automations"}
    };

    auto it = descriptions.find(permission);
    if (it != descriptions.end()) {
        return it->second;
    }
    return "Unknown permission";
}

} // namespace iora

File: sdks/go/permissions.go
// Package iora provides permission types and utilities
package iora

// Permission represents an IORA permission
type Permission string

// All IORA permissions
const (
	// Entity permissions
	PermissionReadEntities    Permission = "ReadEntities"
	PermissionControlEntities Permission = "ControlEntities"
	PermissionCreateEntities  Permission = "CreateEntities"
	PermissionDeleteEntities  Permission = "DeleteEntities"

	// Storage permissions
	PermissionStorageRead   Permission = "StorageRead"
	PermissionStorageWrite  Permission = "StorageWrite"
	PermissionStorageDelete Permission = "StorageDelete"

	// Network permissions
	PermissionNetworkAccess      Permission = "NetworkAccess"
	PermissionNetworkOutbound    Permission = "NetworkOutbound"
	PermissionNetworkInbound     Permission = "NetworkInbound"
	PermissionNetworkScan        Permission = "NetworkScan"
	PermissionNetworkLocalAccess Permission = "NetworkLocalAccess"

	// System permissions
	PermissionSystemInfo    Permission = "SystemInfo"
	PermissionSystemControl Permission = "SystemControl"
	PermissionSystemRestart Permission = "SystemRestart"

	// Database permissions
	PermissionDatabaseRead   Permission = "DatabaseRead"
	PermissionDatabaseWrite  Permission = "DatabaseWrite"
	PermissionDatabaseCreate Permission = "DatabaseCreate"
	PermissionDatabaseDelete Permission = "DatabaseDelete"

	// API and Widget permissions
	PermissionRegisterAPI    Permission = "RegisterApi"
	PermissionCallAPI        Permission = "CallApi"
	PermissionRegisterWidget Permission = "RegisterWidget"
	PermissionControlWidget  Permission = "ControlWidget"

	// Notification permissions
	PermissionSendNotifications Permission = "SendNotifications"
	PermissionReadNotifications Permission = "ReadNotifications"

	// Media permissions
	PermissionCameraAccess     Permission = "CameraAccess"
	PermissionMicrophoneAccess Permission = "MicrophoneAccess"
	PermissionMediaAccess      Permission = "MediaAccess"

	// Plugin and filesystem permissions
	PermissionPluginManager     Permission = "PluginManager"
	PermissionInstallPlugins    Permission = "InstallPlugins"
	PermissionFileSystemRead    Permission = "FileSystemRead"
	PermissionFileSystemWrite   Permission = "FileSystemWrite"
	PermissionFileSystemExecute Permission = "FileSystemExecute"

	// Other permissions
	PermissionLocationAccess Permission = "LocationAccess"
	PermissionAutomations    Permission = "Automations"

	// File sharing permissions (iora-share)
	PermissionFileShareRead   Permission = "FileShareRead"
	PermissionFileShareWrite  Permission = "FileShareWrite"
	PermissionFileShareDelete Permission = "FileShareDelete"
	PermissionFileShareManage Permission = "FileShareManage"

	// Advanced app-only permissions
	PermissionBackupAccess       Permission = "BackupAccess"
	PermissionLogAccess          Permission = "LogAccess"
	PermissionUserManagement     Permission = "UserManagement"
	PermissionSecuritySettings   Permission = "SecuritySettings"
	PermissionNetworkMonitoring  Permission = "NetworkMonitoring"
	PermissionProcessControl     Permission = "ProcessControl"
)

// RiskLevel represents the risk level of a permission
type RiskLevel string

const (
	RiskLevelLow      RiskLevel = "Low"
	RiskLevelMedium   RiskLevel = "Medium"
	RiskLevelHigh     RiskLevel = "High"
	RiskLevelCritical RiskLevel = "Critical"
)

// PermissionCategory defines access level and approval requirements
type PermissionCategory string

const (
	// PluginAllowed - Available to both Apps and Plugins, auto-approved based on manifest
	PermissionCategoryPluginAllowed PermissionCategory = "PluginAllowed"
	// AppOnly - Only available to Apps, auto-approved based on manifest
	PermissionCategoryAppOnly PermissionCategory = "AppOnly"
	// AppOnlyWithConsent - Only available to Apps, requires explicit user consent per installation
	PermissionCategoryAppOnlyWithConsent PermissionCategory = "AppOnlyWithConsent"
)

// PermissionMetadata contains metadata for a permission
type PermissionMetadata struct {
	Permission  Permission
	Category    PermissionCategory
	RiskLevel   RiskLevel
	Description string
}

// GetPermissionRiskLevel returns the risk level for a permission
func GetPermissionRiskLevel(permission Permission) RiskLevel {
	switch permission {
	// Low risk
	case PermissionReadEntities, PermissionStorageRead, PermissionDatabaseRead,
		PermissionSystemInfo, PermissionReadNotifications:
		return RiskLevelLow

	// Critical risk
	case PermissionSystemControl, PermissionSystemRestart, PermissionPluginManager,
		PermissionInstallPlugins, PermissionFileSystemWrite, PermissionFileSystemExecute,
		PermissionNetworkScan, PermissionNetworkLocalAccess:
		return RiskLevelCritical

	// High risk
	case PermissionCreateEntities, PermissionDeleteEntities, PermissionNetworkAccess,
		PermissionNetworkOutbound, PermissionRegisterAPI, PermissionRegisterWidget,
		PermissionCameraAccess, PermissionFileSystemRead:
		return RiskLevelHigh

	// Medium risk (default)
	default:
		return RiskLevelMedium
	}
}

// GetPermissionDescription returns a description for a permission
func GetPermissionDescription(permission Permission) string {
	descriptions := map[Permission]string{
		PermissionReadEntities:       "Read entity states and attributes",
		PermissionControlEntities:    "Control smart home devices",
		PermissionCreateEntities:     "Create new entities",
		PermissionDeleteEntities:     "Delete entities",
		PermissionStorageRead:        "Read from storage",
		PermissionStorageWrite:       "Write to storage",
		PermissionStorageDelete:      "Delete from storage",
		PermissionNetworkAccess:      "Access external networks",
		PermissionNetworkOutbound:    "Make outbound network requests",
		PermissionNetworkInbound:     "Accept inbound network connections",
		PermissionNetworkScan:        "Scan local network",
		PermissionNetworkLocalAccess: "Access local network devices",
		PermissionSystemInfo:         "Read system information",
		PermissionSystemControl:      "Control system settings",
		PermissionSystemRestart:      "Restart IORA system",
		PermissionDatabaseRead:       "Read from database",
		PermissionDatabaseWrite:      "Write to database",
		PermissionDatabaseCreate:     "Create database tables",
		PermissionDatabaseDelete:     "Delete from database",
		PermissionRegisterAPI:        "Register API endpoints",
		PermissionCallAPI:            "Call API endpoints",
		PermissionRegisterWidget:     "Register dashboard widgets",
		PermissionControlWidget:      "Control widgets",
		PermissionSendNotifications:  "Send notifications",
		PermissionReadNotifications:  "Read notifications",
		PermissionCameraAccess:       "Access camera",
		PermissionMicrophoneAccess:   "Access microphone",
		PermissionMediaAccess:        "Access media files",
		PermissionPluginManager:      "Manage plugins",
		PermissionInstallPlugins:     "Install/uninstall plugins",
		PermissionFileSystemRead:     "Read files",
		PermissionFileSystemWrite:    "Write files",
		PermissionFileSystemExecute:  "Execute files",
		PermissionLocationAccess:     "Access location data",
		PermissionAutomations:        "Create and manage automations",
	}

	if desc, ok := descriptions[permission]; ok {
		return desc
	}
	return "Unknown permission"
}
